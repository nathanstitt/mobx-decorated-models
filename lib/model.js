class Box extends BaseModel {
    static attributes = {
        name: null,
        x: 1,
        y: 1,
        someAttribute: 'default value',
    }
    @computed get area() {
        console.log("COMPU");
        return this.x * this.y;
    }
}
