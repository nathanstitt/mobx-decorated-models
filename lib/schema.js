import { observable } from 'mobx';

export default function (model) {
    if (model.$schema) {
        return model.$schema;
    }
    const schema = observable.shallowMap();
    Object.defineProperty(model, '$schema', {
        enumerable: false,
        writable: false,
        configurable: true,
        value: schema,
    });
    return schema;
}
