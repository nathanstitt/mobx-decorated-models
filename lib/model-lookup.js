const ModelsMap = new Map();

function defaultModelLookup({ property, model = property }) {
    if (typeof model === 'function') { return model; }
    return ModelsMap[model] || ModelsMap[property];
}
let modelLookup = defaultModelLookup;

const defaultRememberModel = (model) => {
    ModelsMap[model.identifiedBy || model.name] = model;
};
let rememberModel = defaultRememberModel;

export function registerModel(model) {
    rememberModel(model);
}

export function findModel(model, property) {
    return modelLookup({ model, property });
}

export function rememberModelUsing(fn) {
    rememberModel = fn || defaultRememberModel;
}

export function lookupModelUsing(fn) {
    modelLookup = fn || defaultModelLookup;
}
