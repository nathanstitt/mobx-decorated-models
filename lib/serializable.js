export function isSerializable(model, property) {
    return !!(!model.$nonSerializable || !model.$nonSerializable[property]);
}

export function markNonserializable(model, property) {
    model.$nonSerializable = model.$nonSerializable || {};
    model.$nonSerializable[property] = true;
}
