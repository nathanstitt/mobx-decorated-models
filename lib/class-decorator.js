import {
    serialize, deserialize, createModelSchema, primitive, list, object,
    identifier, getDefaultModelSchema, update,
} from 'serializr';

import { registerModel, findModel } from './model-lookup';
import { isSerializable, markNonserializable } from './serializable';
import getSchema from './schema';

const PendingLookups = [];

function objectSerializer() {
    return {
        serializer(obj, prop, parent) {
            return isSerializable(parent, prop) ? obj : undefined;
        },
        deserializer(json, cb) {
            cb(null, json);
        },
    };
}

function addReference(parentModel, propName, options, cb) {
    const model = findModel(options.model, propName);
    if (model) {
        getDefaultModelSchema(parentModel).props[propName] = cb(model, options, propName);
    } else {
        getDefaultModelSchema(parentModel).props[propName] = objectSerializer();
        PendingLookups.push({ parentModel, propName, options, cb });
    }
}

function getSerializer(options, defaultSerializer) {
    let serializer;
    if (options.type === 'object') {
        serializer = objectSerializer;
    } else if (options.type === 'array') {
        serializer = list;
    } else {
        serializer = defaultSerializer;
    }
    const fns = serializer();
    if (options.writeOnly) {
        return {
            serializer: () => undefined,
            deserializer: fns.deserializer,
        };
    }
    return fns;
}

const Fields = {
    identifier: options => getSerializer(options, identifier),
    field: options => getSerializer(options, primitive),
    session: options =>
        getSerializer(Object.assign({}, options, { writeOnly: true }), primitive),
};

const Associations = {
    hasMany(modelKlass, options) {
        const defaultSerializer = list(object(modelKlass));
        return Object.assign(defaultSerializer, options, {
            model: modelKlass,
        });
    },
    belongsTo(modelKlass, options, propName) {
        const defaultSerializer = object(getDefaultModelSchema(modelKlass));
        return Object.assign(options, {
            model: modelKlass,
            deserializer(value, cb, context) {
                defaultSerializer.deserializer(value, (err, model) => {
                    if (!err && options.inverseOf) {
                        markNonserializable(model, options.inverseOf);
                        model[`${options.inverseOf}_association_name`] = propName;
                        model[options.inverseOf] = context.target;
                    }
                    cb(err, model);
                });
            },
            serializer(belongsTo, name, parent) {
                if (!isSerializable(parent, name)) { return undefined; }
                return defaultSerializer.serializer(belongsTo);
            },
        });
    },
};


const MixedInInstanceMethods = {
    serialize() {
        const schema = getDefaultModelSchema(this.constructor);
        return serialize(schema, this);
    },
    update(json, callback) {
        return update(getDefaultModelSchema(this.constructor), this, json, callback);
    },
};

const MixedInClassMethods = {
    deserialize(json, callback) {
        return deserialize(getDefaultModelSchema(this), json, callback);
    },
};

function decorateModel(model) {
    Object.assign(model.prototype, MixedInInstanceMethods);
    Object.assign(model, MixedInClassMethods);

    const schema = getSchema(model);
    registerModel(model);
    const serializeSchema = {};
    schema.forEach(({ type, options }, name) => {
        if (Fields[type]) {
            serializeSchema[name] = Fields[type](options, name);
        }
    });
    createModelSchema(model, serializeSchema);
    schema.forEach(({ type, options }, name) => {
        if (Associations[type]) {
            addReference(model, name, options, Associations[type]);
        }
    });
    for (let i = PendingLookups.length - 1; i >= 0; i -= 1) {
        const { parentModel, propName, options, cb } = PendingLookups[i];
        const referencedModel = findModel(options.model, propName);
        if (referencedModel) {
            const parentModelSchema = getDefaultModelSchema(parentModel);
            parentModelSchema.props[propName] = cb(referencedModel, options, propName);
            PendingLookups.splice(i, 1);
        }
    }
}

export function modelDecorator(modelOrIdentifier) {
    if (typeof modelOrIdentifier === 'function') {
        modelOrIdentifier.identifiedBy = modelOrIdentifier.name;
        return decorateModel(modelOrIdentifier);
    }
    return (model) => {
        model.identifiedBy = modelOrIdentifier;
        return decorateModel(model);
    };
}

export function unresolvedAssociations() {
    return PendingLookups.map(({ parentModel: model, propName: property }) => (
        { model, property }
    ));
}
