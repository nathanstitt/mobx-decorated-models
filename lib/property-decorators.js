import { observable, intercept } from 'mobx';
import { getDefaultModelSchema } from 'serializr';

import { findModel } from './model-lookup';
import getModelSchema from './schema';
import { markNonserializable } from './serializable'

function addLazyInitializer(target, fn) {
    target.__mobxLazyInitializers.push(fn);
}

function setupModel({
    attrs, modelClass: Klass, array, defaultAttributes,
    inverseOf, parentModel, parentModelProp,
}) {
    if (!attrs || typeof attrs !== 'object') {
        return attrs;
    }
    if (defaultAttributes) {
        if (typeof defaultAttributes === 'function') {
            defaultAttributes = defaultAttributes.call(parentModel, array, parentModel);
        }
        Object.keys(defaultAttributes).forEach((key) => {
            if (!attrs[key]) {
                attrs[key] = defaultAttributes[key];
            }
        });
    }
    if (inverseOf) {
        if (parentModelProp) {
            attrs[`${inverseOf}_association_name`] = parentModelProp;
        }
        attrs[inverseOf] = parentModel;
    }
    const model = (Klass && !(attrs instanceof Klass)) ? new Klass(attrs) : attrs;
    if (inverseOf) {
        markNonserializable(model, inverseOf);
    }
    return model;
}

function onBelongsToSet(change,
                        { modelClass, defaultAttributes, inverseOf,
                          parentModel, parentModelProp }) {
    change.newValue = setupModel({
        attrs: change.newValue, modelClass,
        defaultAttributes, inverseOf, parentModel, parentModelProp,
    });
    return change;
}

function onHasManySet(change,
                      { modelClass, defaultAttributes, inverseOf,
                        parentModel, parentModelProp }) {
    if (change.type !== 'update' || !change.newValue) { return change; }
    const { newValue: array } = change;

    for (let i = 0; i < array.length; i += 1) {
        array[i] = setupModel({
            attrs: array[i], array, modelClass,
            defaultAttributes, inverseOf, parentModel, parentModelProp,
        });
    }

    if (parentModel[parentModelProp]) {
        parentModel[parentModelProp].replace(array);
        return null;
    } else {
        change.newValue = observable.array(array);
        return change;
    }
}

function onCollectionChangeInterceptor(
    { modelClass, model, defaults: defaultAttributes, inverseOf },
    parentModel, parentModelProp,
) {
    return (change) => {
        if (!change.newValue) {
            change.newValue = {};
        }
        if (!modelClass) {
            modelClass = findModel(model, parentModelProp);
        }
        if (change.type === 'splice') {
            for (let i = 0; i < change.added.length; i += 1) {
                change.added[i] = setupModel({
                    attrs: change.added[i], array: change.object, modelClass,
                    defaultAttributes, inverseOf, parentModel, parentModelProp });
            }
        } else if (change.type === 'update') {
            change.newValue = setupModel({
                attrs: change.newValue, array: change.object, modelClass,
                defaultAttributes, inverseOf, parentModel, parentModelProp });
        }
        return change;
    };
}

function buildCollection(options, parentModel, parentModelProp) {
    const ary = observable.array([]);
    if (options.classId || options.model) {
        ary.intercept(onCollectionChangeInterceptor(options, parentModel, parentModelProp));
    }
    return ary;
}

const Initializers = {
    object: () => observable.object({}),
    array: buildCollection,
};

const TypeInitializers = {
    hasMany: buildCollection,
};

function getInitializer(type, options, propName) {
    const fn = TypeInitializers[type] || Initializers[options.type];
    if (!fn) { return undefined; }
    return function () { // eslint-disable-line func-names
        return fn(options, this, propName);
    };
}

function addAttribute(type, target, property, descriptor, options = {}) {
    getModelSchema(target.constructor).set(property, { name: property, type, options });
    const initializer = getInitializer(type, options, property);
    if (initializer) {
        descriptor.initializer = initializer;
    }
    const definition = observable(target, property, descriptor);
    return definition;
}

function buildAttributeDecorator(type, args, attributeAddFn = addAttribute) {
    if ((typeof args[0] === 'object') && args.length === 1) {
        return (target, property, descriptor) =>
            attributeAddFn(type, target, property, descriptor, args[0]);
    }
    return attributeAddFn(type, ...args, {});
}

function interceptingDecorator(interceptingFn) {
    return (type, target, property, descriptor, options = {}) => {
        const decorator = addAttribute(type, target, property, descriptor, options);
        addLazyInitializer(target, (model) => {
            const schema = getDefaultModelSchema(target);
            if (schema && schema.props[property]) {
                const schemaProps = schema.props[property];
                intercept(model, property, (change => interceptingFn(change, {
                    inverseOf:         schemaProps.inverseOf,
                    modelClass:        schemaProps.model,
                    parentModel:       model,
                    parentModelProp:   property,
                    defaultAttributes: schemaProps.defaults,
                })));
            }
        });
        return decorator;
    };
}


const hasMany = (...args) => buildAttributeDecorator(
    'hasMany', args, interceptingDecorator(onHasManySet),
);

const belongsTo = (...args) => buildAttributeDecorator(
    'belongsTo', args, interceptingDecorator(onBelongsToSet),
);

const identifier = (...args) => buildAttributeDecorator('identifier', args);
const field = (...args) => buildAttributeDecorator('field', args);
const session = (...args) => buildAttributeDecorator('session', args);

export { field, session, belongsTo, hasMany, identifier, buildCollection };
