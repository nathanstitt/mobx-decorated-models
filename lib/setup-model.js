import { markNonserializable } from './serializable';

export default function setupModel({
    attrs, modelClass: Klass, array, defaultAttributes,
    inverseOf, parentModel, parentModelProp,
}) {
    if (!attrs) { return attrs; }
    const isObject = ('object' === typeof attrs);
    const canWrite = Object.isExtensible(attrs);

    if (isObject && canWrite) {
        if (defaultAttributes) {
            if ('function' === typeof defaultAttributes) {
                defaultAttributes = defaultAttributes.call(parentModel, array, parentModel);
            }
            Object.keys(defaultAttributes).forEach((key) => {
                if (!attrs[key]) {
                    attrs[key] = defaultAttributes[key];
                }
            });
        }

        if (inverseOf) {
            attrs[inverseOf] = parentModel;
            if (parentModelProp) {
                attrs[`${inverseOf}_association_name`] = parentModelProp;
            }
        }
    }

    const model = (Klass && 'function' === typeof Klass && !(attrs instanceof Klass)) ?
        new Klass(attrs) : attrs;

    if (inverseOf) {
        markNonserializable(model, inverseOf);
    }

    return model;
}
