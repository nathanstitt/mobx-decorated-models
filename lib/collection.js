import { observable } from 'mobx';

import { findModel } from './model-lookup';
import { setupModel } from './setup-model';

function onCollectionChangeInterceptor(
    { modelClass, model, defaults: defaultAttributes, inverseOf },
    parentModel, parentModelProp,
) {
    return (change) => {
        if (!change.newValue) {
            change.newValue = {};
        }
        if (!modelClass) {
            modelClass = findModel(model, parentModelProp);
        }
        // console.log("ARY CHANGE");
        if (change.type === 'splice') {
            for (let i = 0; i < change.added.length; i += 1) {
                change.added[i] = setupModel({
                    attrs: change.added[i],
                    array: change.object,
                    modelClass,
                    defaultAttributes,
                    inverseOf,
                    parentModel,
                    parentModelProp });
            }
        } else if (change.type === 'update') {
            change.newValue = setupModel({
                attrs: change.newValue,
                array: change.object,
                modelClass,
                defaultAttributes,
                inverseOf,
                parentModel,
                parentModelProp });
        }
        return change;
    };
}


export default function createCollection(options, parentModel, parentModelProp) {
    const ary = observable.array([]);
    //console.log("CREATE COL", parentModelProp);
    if (options) {
        if (options.model) {
            ary.intercept(onCollectionChangeInterceptor(options, parentModel, parentModelProp));
        }
        if (options.extend) {
            if ('function' === typeof options.extend){
                options.extend(ary);
            } else {
                Object.assign(ary, options.extend);
            }
        }
    }

    return ary;
}
