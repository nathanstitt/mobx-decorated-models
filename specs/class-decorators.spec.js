import { Container, Box, Ship } from './test-models';
import { unresolvedAssociations } from '../lib/class-decorator';
import { findModel } from '../lib/model-lookup';

describe('Class Decorators', () => {
    it('adds static deserialize method and serialize to prototype', () => {
        const attrs = { id: 42, name: 'TV1', location: 'mid-ship', tags: [], boxes: [
            { id: 1, width: 8, depth: 12, height: 8 },
        ] };
        const container = Container.deserialize(attrs);
        expect(container.boxes).toHaveLength(1);
        expect(container.serialize()).toEqual({
            id: 42, location: 'mid-ship', name: 'TV1', tags: [],
            boxes: [
                { container: undefined, depth: 12, height: 8, id: 1, metadata: {}, width: 8 },
            ],
        });
    });

    it('can deserialize arrays', () => {
        const json = [
            { id: 1, width: 8, depth: 12, height: 8 },
            { id: 2, width: 8, depth: 12, height: 8 },
            { id: 3, width: 8, depth: 12, height: 8 },
        ];
        const boxes = Box.deserialize(json);
        expect(boxes).toHaveLength(3);
        expect(boxes[0].id).toEqual(1);
        expect(boxes[1].id).toEqual(2);
        expect(boxes[2].id).toEqual(3);
    });

    it('can serialize/deserialize nested objects', () => {
        const box = new Box();
        box.update({
            metadata: {
                one: 1,
                two: ['one', 'three'],
                four: { test: true },
            },
        });
        expect(box.serialize().metadata).toEqual(
            expect.objectContaining({
                one: 1,
                four: expect.objectContaining({ test: true }),
                two: expect.arrayContaining(['one', 'three']),
            }),
        );
        expect(box.metadata.one).toEqual(1);
        expect(box.metadata.four.test).toBe(true);
    });

    it('adds update method to prototype', () => {
        const attrs = { id: 2, width: 3, depth: 12, height: 4, container: {
            id: 1, name: '#12', location: 'Building #1',
        } };
        const box = new Box();
        box.update(attrs);
        expect(box.serialize()).toEqual({
            container: { boxes: [], id: 1, location: 'Building #1', name: '#12', tags: [] },
            depth: 12, height: 4, id: 2, metadata: {}, width: 3,
        });
    });

    it('works with belongsTo', () => {
        const box = new Box();
        const container = { id: 1, name: '#12', location: 'Building #1' };
        box.update({ id: 32, width: 3, depth: 12, height: 4, container });
        expect(box.container).toBeInstanceOf(Container);
        expect(box.serialize()).toEqual({
            container: {
                id: 1, boxes: [], tags: [],
                location: 'Building #1', name: '#12',
            },
            depth: 12, height: 4, id: 32, metadata: {}, width: 3,
        });
    });

    it('hasMany', () => {
        const container = new Container({ id: 1, name: 'C23', location: 'z1' });
        container.boxes.push(new Box());
        container.boxes.push(new Box());
        container.boxes[1].width = 4;
        // hasMany sets the inverseOf, but it isn't serialized
        expect(container.boxes[0].container).toBe(container);
        expect(container.serialize()).toEqual({
            id: 1, location: 'z1', name: 'C23', tags: [],
            boxes: [
                { depth: 1, height: 1, metadata: {}, width: 1 },
                { depth: 1, height: 1, metadata: {}, width: 4 },
            ],
        });
    });

    it('uses string identifier with decorator', () => {
        expect(Ship.identifiedBy).toEqual('boat');
        expect(findModel('ship', 'boat')).toBe(Ship)
    });

    it('reports on associations that are not resolved', () => {
        const box = Box.deserialize({ vessel: { id: 1 } });
        expect(box.vessel).toEqual({ id: 1 });
        expect(box.vessel).not.toBeInstanceOf(Ship);

        const pending = unresolvedAssociations();
        expect(pending).toHaveLength(1);
        expect(pending[0].model).toBe(Box);
        expect(pending[0].property).toEqual('vessel');
    });
});
