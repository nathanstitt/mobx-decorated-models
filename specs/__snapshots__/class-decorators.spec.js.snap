// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class Decorators adds static deserialize method and serialize to prototype 1`] = `
Object {
  "boxes": Array [
    Object {
      "depth": 12,
      "height": 8,
      "id": 1,
      "metadata": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@12.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@12",
          "pendingKeys": Map {
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "sides": Array [],
      "vessel": undefined,
      "watercraft": undefined,
      "width": 8,
    },
  ],
  "id": 42,
  "location": "mid-ship",
  "name": "TV1",
  "tags": Array [],
}
`;

exports[`Class Decorators adds update method to prototype 1`] = `
Object {
  "container": Object {
    "boxes": Array [],
    "id": 1,
    "location": "Building #1",
    "name": "#12",
    "tags": Array [],
  },
  "depth": 12,
  "height": 4,
  "id": 2,
  "metadata": Object {
    Symbol(mobx administration): ObservableObjectAdministration$$1 {
      "defaultEnhancer": [Function],
      "keysAtom": Atom$$1 {
        "diffValue": 0,
        "isBeingObserved": false,
        "isPendingUnobservation": false,
        "lastAccessedBy": 0,
        "lowestObserverState": -1,
        "name": "ObservableObject@61.keys",
        "observers": Set {},
      },
      "name": "ObservableObject@61",
      "pendingKeys": Map {
        "$$typeof" => false,
        "nodeType" => false,
        "@@__IMMUTABLE_ITERABLE__@@" => false,
        "@@__IMMUTABLE_RECORD__@@" => false,
        "_isMockFunction" => false,
        "toJSON" => false,
      },
      "proxy": [Circular],
      "target": Object {
        Symbol(mobx administration): [Circular],
      },
      "values": Map {},
    },
  },
  "sides": Array [],
  "vessel": undefined,
  "watercraft": undefined,
  "width": 3,
}
`;

exports[`Class Decorators hasMany 1`] = `
Object {
  "boxes": Array [
    Object {
      "depth": 1,
      "height": 1,
      "id": undefined,
      "metadata": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@96.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@96",
          "pendingKeys": Map {
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "sides": Array [],
      "vessel": undefined,
      "watercraft": undefined,
      "width": 1,
    },
    Object {
      "depth": 1,
      "height": 1,
      "id": undefined,
      "metadata": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@100.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@100",
          "pendingKeys": Map {
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "sides": Array [],
      "vessel": undefined,
      "watercraft": undefined,
      "width": 4,
    },
  ],
  "id": 1,
  "location": "z1",
  "name": "C23",
  "tags": Array [],
}
`;

exports[`Class Decorators works with belongsTo 1`] = `
Object {
  "container": Object {
    "boxes": Array [],
    "id": 1,
    "location": "Building #1",
    "name": "#12",
    "tags": Array [],
  },
  "depth": 12,
  "height": 4,
  "id": 32,
  "metadata": Object {
    Symbol(mobx administration): ObservableObjectAdministration$$1 {
      "defaultEnhancer": [Function],
      "keysAtom": Atom$$1 {
        "diffValue": 0,
        "isBeingObserved": false,
        "isPendingUnobservation": false,
        "lastAccessedBy": 0,
        "lowestObserverState": -1,
        "name": "ObservableObject@70.keys",
        "observers": Set {},
      },
      "name": "ObservableObject@70",
      "pendingKeys": Map {
        "$$typeof" => false,
        "nodeType" => false,
        "@@__IMMUTABLE_ITERABLE__@@" => false,
        "@@__IMMUTABLE_RECORD__@@" => false,
        "_isMockFunction" => false,
        "toJSON" => false,
      },
      "proxy": [Circular],
      "target": Object {
        Symbol(mobx administration): [Circular],
      },
      "values": Map {},
    },
  },
  "sides": Array [],
  "vessel": undefined,
  "watercraft": undefined,
  "width": 3,
}
`;
